lib/Sema/TypeCheckDeclPrimary.cpp

```diff
@@ -235,10 +235,19 @@ static void checkInheritanceClause(
       // AnyObject is not allowed except on protocols.
       if (layout.hasExplicitAnyObject &&
           !isa<ProtocolDecl>(decl)) {
-        decl->diagnose(canHaveSuperclass
-                       ? diag::inheritance_from_non_protocol_or_class
-                       : diag::inheritance_from_non_protocol,
-                       inheritedTy);
+        // Check if the inherited type is exclusively AnyObject
+        // If not, we can be sure of the composition of Protocol(s) with
+        // AnyObject
+        if (!layout.isAnyObject()) {
```

Comment:
I think it's more important to break up the diagnostic in the `else` by making it more specific and using less compiler-jargon-ified language.

Reply:
You mean breaking down the ternary operator to if-else for `canHaveSuperClass`?

Reply:
The intent of the original bug was to fix this particular ternary by providing a more specific message. Its branches are the cause of the confusion here.